### ISSUE TITLE
slow fallback array `copyto!` in Julia 1.7.0

### ISSUE THREAD

**Issue by johnnychen94** [JuliaLang/julia#43287]
It looks like a potential type inference issue

```julia
using BenchmarkTools
X = rand(ComplexF32, 64, 64);

dst = reinterpret(reshape, Float32, X);
src = copy(dst);

@btime copyto!($dst, $src);
# 1.6.4: 42.120 μs (0 allocations: 0 bytes)
# 1.7.0: 800.737 μs (32768 allocations: 1.12 MiB)
# 1.8.0-DEV.1090: 829.629 μs (32768 allocations: 1.12 MiB)
```

Profiling the code shows that it's hitting the fallback dual-iterator implementation version and a slow Tuple iterator

https://github.com/JuliaLang/julia/blob/d16f4806e9389dbc92c463efc5b96f67a7aebf22/base/abstractarray.jl#L1041-L1047
https://github.com/JuliaLang/julia/blob/d16f4806e9389dbc92c463efc5b96f67a7aebf22/base/tuple.jl#L88

![image](https://user-images.githubusercontent.com/8684355/144288905-59917920-9a32-48fc-9ce9-24838450fed2.png)

Downstream JuliaImages issues: https://github.com/JuliaImages/Images.jl/issues/992 https://github.com/JuliaImages/ImageCore.jl/issues/174 (cc: @etibarg @timholy)

**Comment by N5N3** [comment]
`@code_warntype` gives no warning.
But `@code_typed` shows that:
```julia
9 ── %70  = %new(Base.SCartesianIndex2{2}, 1, %65)::Base.SCartesianIndex2{2}
│    %71  = Core.tuple(1, %65, %66)::Tuple{Int64, Int64, Int64}
│    %72  = Core.tuple(%70, %71)::Tuple{Base.SCartesianIndex2{2}, Tuple{Int64, Int64, Int64}}
└───        goto #10
10 ┄ %74  = φ (#9 => %70)::Union{Base.SCartesianIndex2{2}, Tuple{Int64, Int64, Int64}}
│    %75  = φ (#9 => %71)::Union{Base.SCartesianIndex2{2}, Tuple{Int64, Int64, Int64}}
│    %76  = φ (#8 => %68, #9 => %72)::Union{Nothing, Tuple{Base.SCartesianIndex2{2}, Tuple{Int64, Int64, Int64}}}
│    %77  = Base.bitcast(UInt64, 1)::UInt64
│    %78  = Base.sub_int(%77, 0x0000000000000001)::UInt64
│    %79  = Base.arraylen(src)::Int64
│    %80  = Base.sle_int(0, %79)::Bool
│    %81  = Base.bitcast(UInt64, %79)::UInt64
│    %82  = Base.ult_int(%78, %81)::Bool
│    %83  = Base.and_int(%80, %82)::Bool
└───        goto #12 if not %83
```
Looks like we messes up at `%74` and `%75`

**Comment by timholy** [comment]
`git bisect` blames 1f21f2d6feacaa5bf3f7ea212c64e98bd66fd265

**Comment by aviatesk** [comment]
Thanks for the bisect. Will be fixed by #43347.
